From c21a962d1abb50edde003dc37625d980ca4e9c74 Mon Sep 17 00:00:00 2001
From: Jason Wessel <jason.wessel@windriver.com>
Date: Thu, 9 Aug 2012 12:35:10 -0500
Subject: [PATCH 19/29] mips,kdb: Add software single stepping support

Add the 's' command to handle software single stepping in KDB on MIPS.

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
---
 arch/mips/kernel/kgdb.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/arch/mips/kernel/kgdb.c b/arch/mips/kernel/kgdb.c
index fcaac2f..4868861 100644
--- a/arch/mips/kernel/kgdb.c
+++ b/arch/mips/kernel/kgdb.c
@@ -338,7 +338,7 @@ static struct notifier_block kgdb_notifier = {
 };
 
 /*
- * Handle the 'c' command
+ * Handle the 's' and 'c' commands
  */
 int kgdb_arch_handle_exception(int vector, int signo, int err_code,
 			       char *remcom_in_buffer, char *remcom_out_buffer,
@@ -346,14 +346,20 @@ int kgdb_arch_handle_exception(int vector, int signo, int err_code,
 {
 	char *ptr;
 	unsigned long address;
+	int cpu = smp_processor_id();
 
 	switch (remcom_in_buffer[0]) {
+	case 's':
 	case 'c':
 		/* handle the optional parameter */
 		ptr = &remcom_in_buffer[1];
 		if (kgdb_hex2long(&ptr, &address))
 			regs->cp0_epc = address;
 
+		atomic_set(&kgdb_cpu_doing_single_step, -1);
+		if (remcom_in_buffer[0] == 's')
+			atomic_set(&kgdb_cpu_doing_single_step, cpu);
+
 		return 0;
 	}
 
-- 
1.8.4.93.g57e4c17

